### ISSUE TITLE
istringstream fail on parse when fractional seconds and timezone offset are used together

### ISSUE THREAD

**Issue by eyepatchParrot** [HowardHinnant/date#511]
I'd like to parse a timestamp with optional fractional seconds and a timezone offset. Each works individually, but they don't work together.

    bool
    is_parse_ok(const std::string& str, const char * fmt) {
        std::istringstream in(str);
        date::sys_seconds tp;
        in >> date::parse(fmt, tp);
        return !in.fail();
    }
    int main() {
        std::cout << is_parse_ok("2015-08-27T11:31:40.533", "%FT%T") << '\n'
                  << is_parse_ok("2015-08-27T11:31:40+0100", "%FT%T%z") << '\n'
                  << is_parse_ok("2015-08-27T11:31:40.533+0100", "%FT%T%z") << std::endl;
    }

**Comment by HowardHinnant** [comment]
I'm not sure the first is working either.  Are you sure you want to try to parse optional fractional seconds into a time point with seconds precision?

**Comment by eyepatchParrot** [comment]
You're right. "%FT%H:%M:%6S%z" does what I was looking for. It's a little surprising though that I didn't come across this in any of the recipes. It's also not mentioned in %T, and I misunderstood %S the first few times I read it.

"If not specified, the default is 2 if the input time has a precision convertible to seconds. Otherwise the default width is determined by the decimal precision of the input and the field is interpreted as a long double in a fixed format. "

My understanding of this was that if the input time has a precision which is not convertible to seconds, then it uses decimal precision. I see that what might have been meant was that decimal precision is considered only if the width is specified. To be precise, my confusion was whether 'otherwise' was referring to whether it's convertible to seconds or whether it's specified.

**Comment by HowardHinnant** [comment]
I've modified your code slightly to print out the time point if the parse succeeds, and otherwise print out fail.  Here's your original code with that modification:
```c++
#include "date/date.h"
#include <iostream>
#include <sstream>

using TP = date::sys_seconds;

struct parse_return
{
    TP tp;
    bool is_ok;

    friend
    std::ostream&
    operator<<(std::ostream& os, parse_return const& pr)
    {
        using date::operator<<;
        if (pr.is_ok)
            return os << pr.tp;
        return os << "fail";
    }
};

parse_return
is_parse_ok(const std::string& str, const char * fmt) {
    std::istringstream in(str);
    TP tp;
    in >> date::parse(fmt, tp);
    return {tp, !in.fail()};
}

int main() {
    std::cout << is_parse_ok("2015-08-27T11:31:40.533", "%FT%T") << '\n'
              << is_parse_ok("2015-08-27T11:31:40+0100", "%FT%T%z") << '\n'
              << is_parse_ok("2015-08-27T11:31:40.533+0100", "%FT%T%z") << std::endl;
}
```
And the output is:
```
2015-08-27 11:31:40
2015-08-27 10:31:40
fail
```
Note that the first parse did not capture the ".533".  Indeed, it didn't even try to parse it.  So subsequent parses on the same stream are likely to fail unless they know to skip over the ".533".

Here is the output using  "%FT%H:%M:%6S%z" instead for the 2nd and 3rd parse:
```
2015-08-27 11:31:40
2015-08-27 10:31:40
2015-08-27 10:31:41
```
The last one is rounding to the nearest second.  I don't know if that's what you want.

I recommend parsing into a `date::sys_time<std::chrono::milliseconds>`:

```c++
using TP = date::sys_time<std::chrono::milliseconds>;
```

Using "%FT%T%z" for the 2nd and 3rd parse:

```
2015-08-27 11:31:40.533
2015-08-27 10:31:40.000
2015-08-27 10:31:40.533
```

This captures the full precision of each input.  Parsing into `std::chrono::system_clock::time_point` will also work.
```
2015-08-27 11:31:40.533000
2015-08-27 10:31:40.000000
2015-08-27 10:31:40.533000
```
The problem with parsing into `sys_seconds` (which is a type alias for `sys_time<seconds>`, which is a type alias for `time_point<system_clock, seconds>`), is that it has no place to put the fractional seconds, and therefore stops parsing as soon as it reaches the decimal point (or has parsed two decimal digits).  By parsing into a finer precision `time_point`, you ask to parse the decimal point (if it exists) and fractional seconds digits (if they exist).

If you _dont_ _want_ the extra precision after parsing the fractional seconds, I still recommend doing the parse with a higher precision `time_point`, and then using `floor`, `ceil`, `round` or `time_point_cast` to truncate that higher-preicsion `time_point` into `sys_seconds`.
